macaroonbakery/checkers/checkers.py

```diff
@@ -1,53 +1,231 @@
 # Copyright 2017 Canonical Ltd.
 # Licensed under the LGPLv3, see LICENCE file for details.
 
-import collections
-import datetime
+from datetime import datetime
 
-import rfc3339
+from pyrfc3339 import parse
+import pytz
 
-from macaroonbakery.checkers.namespace import STD_NAMESPACE
+from macaroonbakery.checkers.declared import DECLARED_KEY
+from macaroonbakery.checkers.time import TIME_KEY
+from macaroonbakery.checkers.operation import OP_KEY
+from macaroonbakery.checkers.namespace import Namespace
+from macaroonbakery.checkers.caveat import parse_caveat
+from macaroonbakery.checkers import (
+    STD_NAMESPACE, COND_DECLARED, COND_ALLOW, COND_DENY, COND_ERROR,
+    COND_TIME_BEFORE
+)
+from macaroonbakery.checkers.utils import condition_with_prefix
 
 
-class Caveat(collections.namedtuple('Caveat', 'condition location namespace')):
-    '''Represents a condition that must be true for a check to complete
-    successfully.
+class Checker(object):
+    ''' Holds a set of checkers for first party caveats.
+    '''
+
+    def __init__(self, ns=None, include_std_checkers=True):
+        if ns is None:
+            ns = Namespace()
+        self._ns = ns
+        self._checkers = {}
+        if include_std_checkers:
+            self.register_std()
+
+    def check_first_party_caveat(self, ctx, cav):
+        ''' Checks the caveat against all registered caveats conditions.
+        :return: error message if any or None
+        '''
+        try:
+            cond, arg = parse_caveat(cav)
+        except ValueError as ex:
+            # If we can't parse it, perhaps it's in some other format,
+            # return a not-recognised error.
+            return 'cannot parse caveat "{}": {}'.format(cav, ex.args[0])
+        checker = self._checkers.get(cond)
+        if checker is None:
+            return 'caveat "{}" not satisfied: caveat not recognized'.format(
+                cav)
+        err = checker.check(ctx, cond, arg)
+        if err is not None:
+            return 'caveat "{}" not satisfied: {}'.format(cav, err)
+
+    def namespace(self):
+        ''' Returns the namespace associated with the Checker.
+        '''
+        return self._ns
+
+    def info(self):
+        ''' Returns information on all the registered checkers.
+
+        Sorted by namespace and then name
+        :returns a list of CheckerInfo
+        '''
+        return sorted(self._checkers.values(), key=lambda x: (x._ns, x._name))
+
+    def register(self, cond, uri, check):
+        ''' Registers the given condition in the given namespace URI
+        to be checked with the given check function.
```

Comment:
Please document the expected types of the arguments and the contract of the check function.